check_patroni/types.py

```diff
@@ -27,27 +36,58 @@ class Parameters:
 class PatroniResource(nagiosplugin.Resource):
     conn_info: ConnectionInfo
 
-    def rest_api(
-        self: "PatroniResource", service: str
-    ) -> urllib3.response.HTTPResponse:
+    def rest_api(self: "PatroniResource", service: str) -> Any:
         """Try to connect to all the provided endpoints for the requested service"""
         for endpoint in self.conn_info.endpoints:
             try:
+                cert: Optional[Union[Tuple[str, str], str]] = None
+                verify: Optional[Union[str, bool]] = None
                 if endpoint[:5] == "https":
-                    pool = urllib3.PoolManager(
-                        cert_reqs="CERT_REQUIRED",
-                        cert_file=self.conn_info.cert_file,
-                        key_file=self.conn_info.key_file,
-                        ca_certs=self.conn_info.ca_cert,
+                    if (
+                        self.conn_info.cert_file is not None
+                        and self.conn_info.key_file is not None  # noqa W503
+                    ):
+                        # we provide a certificate and a private key
+                        cert = (self.conn_info.cert_file, self.conn_info.key_file)
+                    elif (
+                        self.conn_info.cert_file is not None
+                        and self.conn_info.key_file is None  # noqa W503
+                    ):
+                        # we provide a pem file with the private key and the certificate
+                        cert = self.conn_info.cert_file
+
+                    if self.conn_info.ca_cert is not None:
+                        # if cert is not None: this is the CA certificate
+                        # otherwise this is a ca bundle with root certificate
+                        # then some optional intermediate certificate and finally
+                        # the cerver certificate to validate the certification chain
+                        verify = self.conn_info.ca_cert
+                    else:
+                        if cert is None:
+                            # if cert is None we want to bypass https verification,
+                            # this is in secure and should be avoided for production use
+                            verify = False
+
+                _log.debug(
+                    f"Trying to connect to {endpoint}/{service} with cert: {cert} verify: {verify}"
+                )
+
+                # We want to disable warnings such as InsecureRequestWarning or
+                # SubjectAltNameWarning. It breaks the output of the script, which
+                # is a problem for nagios.
+                # requests uses urllib3 under the hood, we use it to disable warnings,
+                urllib3.disable_warnings()
+
+                r = requests.get(f"{endpoint}/{service}", verify=verify, cert=cert)
+                _log.debug(f"api call status: {r.status_code}")
+                _log.debug(f"api call data: {r.text}")
```

Comment:
Instead of logging yourself, you can raise the log level of `urllib3` and you'll get its log messages:
```python
logging.getLogger("urllib3").setLevel(logging.DEBUG)
```

Reply:
It's not working. But I have had weird stuff append to me with loging & nagiosplugin.

Could this be the reason? https://github.com/mpounsett/nagiosplugin/blob/develop/nagiosplugin/runtime.py#L81

Reply:
And the setup for `urllib3` should be done close to global logging configuration.

Reply:
I tried adding this line at the beginning of `cli.py` before:
```
logging.basicConfig(format='%(levelname)s - %(message)s')
```

But I got double logging:
```
$ check_patroni -vvv -e https://p1:8008 cluster_node_count
DEBUG - Trying to connect to https://p1:8008/cluster with cert: None verify: False
/home/benoit/git/dalibo/check_patroni/.venv/lib64/python3.11/site-packages/urllib3/connectionpool.py:1045: InsecureRequestWarning: Unverified HTTPS request is being made to host 'p1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
DEBUG - api call status: 200
DEBUG - api call data: {"members": [{"name": "p1", "role": "leader", "state": "running", "api_url": "https://10.20.30.51:8008/patroni", "host": "10.20.30.51", "port": 5432, "timeline": 6}, {"name": "p2", "role": "replica", "state": "running", "api_url": "https://10.20.30.52:8008/patroni", "host": "10.20.30.52", "port": 5432, "timeline": 6, "lag": 0}]}
CLUSTERNODECOUNT OK - members is 2
Trying to connect to https://p1:8008/cluster with cert: None verify: False
api call status: 200
api call data: {"members": [{"name": "p1", "role": "leader", "state": "running", "api_url": "https://10.20.30.51:8008/patroni", "host": "10.20.30.51", "port": 5432, "timeline": 6}, {"name": "p2", "role": "replica", "state": "running", "api_url": "https://10.20.30.52:8008/patroni", "host": "10.20.30.52", "port": 5432, "timeline": 6, "lag": 0}]}
| members=2 role_leader=1 role_replica=1 state_running=2
```

I thought it was because nagiosplugin does black magic with the rootlogger in my back.

Reply:
done in https://github.com/dalibo/check_patroni/pull/22/commits/796ca8e5765cd66c50dfb78c33e30cd2eaa379b4